CONTRIBUTING.md

```diff
@@ -0,0 +1,168 @@
+- `F841: local variable is assigned to but never used`
+    - Must ignore because gt4py stencils return outputs in place (no return statement)
+    - Can avoid if all stencil assignments use square brackets on the left
+        - `out[0, 0, 0]` or `out[idx]/out[curr]/out[something]`
+
+Flake8 rules not listed will be enforced unless we find a need not to enforce them.
+Since documentation does go out of date, please consult the entry in `setup.cfg` for the most up-to-date requirements and update this document if you notice a difference.
+
+## Style
+
+The first version of the dycore was written with minimal metadata and typing, motivated primarily by matching the regression data produced by the Fortran version of the code using the numpy backend.
+We are now actively refactoring, moving code that still does computations in Python into GT4py stencils and merging stencils together with the introduction of enabling features in GT4py (such as regions).
+While we do that, clarifying the operation of the model and what the variables are will both help make the model easier to read and reduce errors as we move around long lists of argument variables.
+
+Specifically, we want to start adding the following where appropriate:
+- Type hints on Python functions (see [`fv3core/utils/typing.py`](https://github.com/VulcanClimateModeling/fv3core/blob/master/fv3core/utils/typing.py) and below)
+- More descriptive types on stencil definitions
+- Docstrings on outward facing Python functions: describe what methods are doing, describe the intent (*in*, *out*, or *inout*) of the function arguments
+
+### Docstrings
+These should aid us in refactoring and understanding what a function is doing. If it is not completely understood yet what is happening, please consult other team members and GFDL as appropriate. If we still cannot tell what is happening, you can write what is known along with a `TODOC` to indicate it is incomplete.
+For example:
+
+```python
+def stencil(...):
+    """This is a short description that fits on one line.
+
+    Here is a longer explanation of the assumptions and why this exists.
+
+    TODOC: Why does ke_c need to be updated here?
+
+    Args:
+        uc: x-velocity on C-grid (inout)
+        vc: y-velocity on C-grid (inout)
+        vort_c: Vorticity on C-grid (inout)
+        ke_c: kinetic energy on C-grid (inout)
+        v: y-velocity on D-grid (inout)
+        u: x-velocity on D-grid (inout)
+        dt2: timestep (in)
+
+    """
+```
+
+
+### Type hinting Python functions
+These should mostly be lightweight workflow wrappers calling gt4py stencils, though currently exceptions exist where Python code does computations on data fields.
+
+Original convention is:
+```python
+def compute(var1, var2, var3, param1, param2, param3):
+```
+
+Order of arguments does not actually matter, but generally follows the convention of listing 3d fields first, followed by parameters, as is required by gt4py stencil functions.
```

Comment:
I first read this (inaccurately) as "gt4py stencils can take their arguments in any order". There might be a more precise way to say it than "order does not actually matter".

Reply:
ah, but it does matter for stencils, it's python functions that it doesn't. and could definitely be said differently.